src/Corax/Queries/SortingMultiMatch.cs

```diff
@@ -0,0 +1,498 @@
+                }
+                else if (typeof(TW) == typeof(NumericalItem<long>))
+                {
+                    return comparer.CompareNumerical(((NumericalItem<long>)(object)ix.Value).Value, ((NumericalItem<long>)(object)iy.Value).Value);
+                }
+                else if (typeof(TW) == typeof(NumericalItem<double>))
+                {
+                    return comparer.CompareNumerical(((NumericalItem<double>)(object)ix.Value).Value, ((NumericalItem<double>)(object)iy.Value).Value);
+                }
+                return -1;
+            }
+        }
+
+        [SkipLocalsInit]
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private int Fill<W>(Span<long> matches) where W : struct
+        {
+            // Important: If you are going to request a massive take like 20K you need to pass at least a 20K size buffer to work with.
+            //            The rationale for such behavior is that sorting has to find among the candidates the order between elements,
+            //            and it can't do so without checking every single element found. If you fail to do so, your results may not be
+            //            correct. 
+            Debug.Assert(_take <= matches.Length);
+
+            int totalMatches = _inner.Fill(matches);
+            if (totalMatches == 0)
+                return 0;
+
+            int take = _take <= 0 ? matches.Length : Math.Min(matches.Length, _take);
+            TotalResults += totalMatches;
+
+            var matchesKeysHolder = QueryContext.MatchesPool.Rent(2 * Unsafe.SizeOf<MultiMatchComparer<TComparer1, W>.Item>() * matches.Length);
+            var matchesKeysSpan = MemoryMarshal.Cast<byte, MultiMatchComparer<TComparer1, W>.Item>(matchesKeysHolder);
+
+            // PERF: We want to avoid to share cache lines, that's why the second array will move toward the end of the array. 
+            var matchesKeys = matchesKeysSpan[0..matches.Length];
+            var bKeys = matchesKeysSpan[^matches.Length..];
+
+
+            var searcher = _searcher;
+            var fieldId = _comparer1.FieldId;
+            var comparer = new MultiMatchComparer<TComparer1, W>(this);            
+            for (int i = 0; i < totalMatches; i++)
+            {
+                UnmanagedSpan matchIndexEntry = searcher.GetIndexEntryPointer(matches[i]);
+                var read = Get(new IndexEntryReader(matchIndexEntry), fieldId, matches[i], out matchesKeys[i].Value);
+                matchesKeys[i].Key = read ? matches[i] : -matches[i];
+                matchesKeys[i].Entry = matchIndexEntry;
+            }
+
+            // We sort the first batch. That will also mean that we will sort the indexes too. 
+            var sorter = new Sorter<MultiMatchComparer<TComparer1, W>.Item, long, MultiMatchComparer<TComparer1, W>>(comparer);
+            sorter.Sort(matchesKeys[0..totalMatches], matches);
+
+            Span<long> bValues = stackalloc long[matches.Length];
```

Comment:
This is probably wrong. we *are* going to get `take int.MaxValue` so you cannot use this.

Reply:
You can. That value comes from the `Fill(...)` the `take` happens after during the algorithm execution.

Reply:
Unless I miss something, if `take` is MaxValue, then `matches` must be at least as long.
If you then also have enough matches, you are going to overflow the stack

Reply:
Only if we pass a big enough matches, yeah, we should guard against that anyways...

Reply:
`from Users where Active = true` is a query that would need that, no?

Reply:
Unless you are willing to pass a multimegabyte buffer to that kind of query. Nope.

Reply:
?? That _is_ a query that a user will send, for sure.

Reply:
I know, but at some point you will have to put a limit... via say 'streaming' or multi-step or something... so that it is bounded by the batch size. The only one that really could need to handle that is sorting. Which we will cross that road when we need to.

Reply:
Not really. This is a real scenario, and we need to give an an answer to that.
Not right away, but when you use `stackalloc` with an unbounded value, we have to ensure that this won't break for the future.

Reply:
Right now, with `Lucene` in this case, it allocates an `int[]` with `Math.Min(take, numOfDocs)` to hold the data.
The `take` is often MaxValue, and the `numOfDocs` depends on the actual size of the index. That can be another limit, but we *are* going to have large indexes, so that is a 1st class worry right now

Reply:
Yes, that's why we need to think in terms of multi-step when in the face of sorting. So far, it is the only operation that requires you to get all the matches in 1 place. But for that we need to know how many potential matches to have.

Reply:
If you need a sorted query over large number of results, you'll just pay the memory cost for that. This is reasonable to expect, as such, we'll just allocate that amount as needed.

Reply:
That settles the issue then, because I was going to ask for that memory outside of the stack anyways when it goes over 16K elements...